Feature/zebedee collection content

CarlHembrough opened the pull request:
- Add UpdateCollectionContent function
- Refactor HttpClient to use dp-net HTTP client
- Add test coverage to existing CreateCollection and DeleteCollection functions
- move requestObject function from HttpClient to zebedeeClient, making it consistent with the other helper functions.

CarlHembrough committed:
Add additional Zebedee collection functions

- add content to collection
- delete content from collection
- complete content
- review content
- approve collection
- unlock collection

CarlHembrough committed:
Change HttpClient interface to match dp-net HTTP client

- move requestObject function from HttpClient to zebedeeClient

CarlHembrough committed:
Add test coverage for delete collection function

CarlHembrough committed:
Check response bool value from when updating collection content

- add test coverage for update collection content

CarlHembrough committed:
Remove functions not yet covered by tests

CarlHembrough committed:
Use constants to reduce noise in tests

CarlHembrough committed:
Use constants for Florence header key

CarlHembrough committed:
Use constants for error message assertion

daiLlew commented on zebedee/collection.go:
```diff
@@ -113,3 +115,38 @@ func (z *zebedeeClient) UpdateCollection(s Session, desc CollectionDescription)
+	}
+
+	req.Header.Set("content-type", "application/json")
+	req.Header.Set(request.FlorenceHeaderKey, s.ID)
```
There is  `z.newAuthenticatedRequest()` to create requests with the auth header.

CarlHembrough replied:
I had initially tried to use this, but then didn't because of the io.reader usage - see point below. I have used `z.newAuthenticatedRequest()` for all other functions that follow this PR

CarlHembrough replied:
This function is now using `newAuthenticatedRequest`

daiLlew commented on zebedee/collection.go:
```diff
@@ -113,3 +115,38 @@ func (z *zebedeeClient) UpdateCollection(s Session, desc CollectionDescription)
+func (z *zebedeeClient) UpdateCollectionContent(
+	s Session,
+	id, contentUri string,
+	content io.Reader,
```
1.  Does this need to be an `io.Reader`? I know the page/content types that are written to this endpoint are varied so a single type won't work but could we use `interface{}` instead?  I think that would offer the same flexibility but also allow you to reuse the `z.newAuthenticatedRequest()` func to create the requests. 

2. What are the overwriteExisting, recursive and validate JSON defaults? If there is a consistent default could we apply that internally within this method and keep the interface simple by reducing the number of params..

I'm aware I am nit picking but I want to keep it clean and simple where possible.

CarlHembrough replied:
Fair points
> 1. Does this need to be an `io.Reader`? I know the page/content types that are written to this endpoint are varied so a single type won't work but could we use `interface{}` instead?  I think that would offer the same flexibility but also allow you to reuse the `z.newAuthenticatedRequest()` func to create the requests.

In the case of the collection API I don't think the JSON content will be unmarshalled before calling Zebedee - it will just be passed through the API. This is the reasoning behind having the reader, as it can pass the reader from the body of the request without having to unmarshall. 🤔 

> 2. What are the overwriteExisting, recursive and validate JSON defaults? If there is a consistent default could we apply that internally within this method and keep the interface simple by reducing the number of params..

I added some comments to the function header describing their purpose. I debated what to do with these - they are a side effect of the content endpoint doing too much. I had considered setting defaults, though Florence will use different values depending on the content type. I get your point about the params being too much - I guess another option would be to use this function within the package, and expose multiple content function for the different variations of the parameters. Trouble is it's not easy to decipher what they are 🤯

CarlHembrough replied:
I have reworked the interface to
 - use the `interface{}` instead of `io.Reader`. It's seems likely that the JSON will be unmarshalled to some extent to it seems the right solution 👍 
 - I have applied the default values for these flags internally. If there are variations required, then the decision can be made how to expose the variations.

CarlHembrough committed:
Refactor UpdateCollectionContent content type

- was using io.Reader, but changed to use interface{} for consistency

CarlHembrough committed:
Use default flags within UpdateCollectionContent

Zebedee allows a number of flags to be passed alongside the request to change the semantics of the content update. It is not yet understood where/if the flags will be required, so they have been set to the defaults for now to minimise the parameters.
